### PR TITLE
Role security expression & fix comments

### PR DESCRIPTION
**Pull request by rimnvd**

### CODE REVIEW

**Comment by IP-13** [comment]
|Overall Project|73.22% **`-5.71%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|18.31%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[PrivilegeService.java](https://github.com/itmo-events-app/event-app/blob/3d9edd1f1db2a3fc9d2c2c9dde85199105864526/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FPrivilegeService.java)|100%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/3d9edd1f1db2a3fc9d2c2c9dde85199105864526/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|90.63%|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/3d9edd1f1db2a3fc9d2c2c9dde85199105864526/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|67.52% **`-12.79%`**|:x:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/3d9edd1f1db2a3fc9d2c2c9dde85199105864526/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|25% **`-28.62%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/3d9edd1f1db2a3fc9d2c2c9dde85199105864526/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|6.09% **`-4.06%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/3d9edd1f1db2a3fc9d2c2c9dde85199105864526/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|5.52% **`-94.48%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|72.95% **`-6.44%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|26.58%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[PrivilegeService.java](https://github.com/itmo-events-app/event-app/blob/53705512827ce055ad67603d98e1a36d222c55de/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FPrivilegeService.java)|100%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/53705512827ce055ad67603d98e1a36d222c55de/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|89.02%|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/53705512827ce055ad67603d98e1a36d222c55de/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|67.18% **`-15.52%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/53705512827ce055ad67603d98e1a36d222c55de/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|22.82% **`-34.83%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/53705512827ce055ad67603d98e1a36d222c55de/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|6.09% **`-4.06%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/53705512827ce055ad67603d98e1a36d222c55de/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|5.52% **`-94.48%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|73.7% **`-6.26%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|26.58%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[PrivilegeService.java](https://github.com/itmo-events-app/event-app/blob/64331b5dff437d911c944514a0042e75215aa6b7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FPrivilegeService.java)|100%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/64331b5dff437d911c944514a0042e75215aa6b7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|89.02%|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/64331b5dff437d911c944514a0042e75215aa6b7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|67.18% **`-15.52%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/64331b5dff437d911c944514a0042e75215aa6b7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|24.41% **`-34.12%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/64331b5dff437d911c944514a0042e75215aa6b7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.06% **`-3.85%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/64331b5dff437d911c944514a0042e75215aa6b7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|5.52% **`-94.48%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|75.03% **`-6%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|25.6%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[PrivilegeService.java](https://github.com/itmo-events-app/event-app/blob/aa90cf88334eb3c7639d80414a74bcab75c4bee4/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FPrivilegeService.java)|100%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/aa90cf88334eb3c7639d80414a74bcab75c4bee4/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|89.02%|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/aa90cf88334eb3c7639d80414a74bcab75c4bee4/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|67.18% **`-15.52%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/aa90cf88334eb3c7639d80414a74bcab75c4bee4/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|29.45% **`-31.49%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/aa90cf88334eb3c7639d80414a74bcab75c4bee4/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.06% **`-3.85%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/aa90cf88334eb3c7639d80414a74bcab75c4bee4/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|5.52% **`-94.48%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|75.03% **`-6.11%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|25.27%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[PrivilegeService.java](https://github.com/itmo-events-app/event-app/blob/3a61b21fb6f0c93d64dd31b52e231e2f6bb0e444/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FPrivilegeService.java)|100%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/3a61b21fb6f0c93d64dd31b52e231e2f6bb0e444/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|89.02% **`-3.66%`**|:x:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/3a61b21fb6f0c93d64dd31b52e231e2f6bb0e444/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|67.18% **`-15.52%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/3a61b21fb6f0c93d64dd31b52e231e2f6bb0e444/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|29.45% **`-31.49%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/3a61b21fb6f0c93d64dd31b52e231e2f6bb0e444/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.06% **`-3.85%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/3a61b21fb6f0c93d64dd31b52e231e2f6bb0e444/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|5.52% **`-94.48%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|75.03% **`-6.11%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|26.23%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[PrivilegeService.java](https://github.com/itmo-events-app/event-app/blob/1a6afbde96f9a5f6149fa3508c515983239127b8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FPrivilegeService.java)|100%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/1a6afbde96f9a5f6149fa3508c515983239127b8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|89.02% **`-3.66%`**|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/1a6afbde96f9a5f6149fa3508c515983239127b8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|67.18% **`-15.52%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/1a6afbde96f9a5f6149fa3508c515983239127b8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|29.45% **`-31.49%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/1a6afbde96f9a5f6149fa3508c515983239127b8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.06% **`-3.85%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/1a6afbde96f9a5f6149fa3508c515983239127b8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|5.52% **`-94.48%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|75.03% **`-6.11%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|26.23%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[PrivilegeService.java](https://github.com/itmo-events-app/event-app/blob/e27a54719aed6cbcb919e2a0c25f126c525e861e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FPrivilegeService.java)|100%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/e27a54719aed6cbcb919e2a0c25f126c525e861e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|89.02% **`-3.66%`**|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/e27a54719aed6cbcb919e2a0c25f126c525e861e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|67.18% **`-15.52%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/e27a54719aed6cbcb919e2a0c25f126c525e861e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|29.45% **`-31.49%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/e27a54719aed6cbcb919e2a0c25f126c525e861e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.06% **`-3.85%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/e27a54719aed6cbcb919e2a0c25f126c525e861e/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|5.52% **`-94.48%`**|:x:|

File: src/main/java/org/itmo/eventapp/main/controller/RoleController.java [hunk 1]
```diff
@@ -84,24 +108,28 @@ public ResponseEntity<RoleResponse> editRole(@Positive(message = "Парамет
...
     public ResponseEntity<List<PrivilegeResponse>> getSystemPrivileges() {
         return ResponseEntity.ok(PrivilegeMapper.privilegesToPrivilegeResponseList(privilegeService.getPrivilegeByType(PrivilegeType.SYSTEM)));
     }
 
     @Operation(summary = "Получение списка организационных привилегий")
+    @PreAuthorize("@roleSecurityExpression.canCreateRole() or @roleSecurityExpression.canEditRole()")
     @GetMapping("/organizational-privileges")
     public ResponseEntity<List<PrivilegeResponse>> getOrganizationalPrivileges() {
         return ResponseEntity.ok(PrivilegeMapper.privilegesToPrivilegeResponseList(privilegeService.getPrivilegeByType(PrivilegeType.EVENT)));
     }
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/controller/RoleController.java:129]: Если тут одинаковые роли, может сделать одним методом с параметром privilegeType?

Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/controller/RoleController.java:129]: Если тут одинаковые роли, может сделать одним методом с параметром privilegeType?

File: src/main/java/org/itmo/eventapp/main/controller/RoleController.java [hunk 2]
```diff
@@ -121,6 +150,7 @@ public ResponseEntity<Void> assignOrganizerRole(
...
 
     @Operation(summary = "Назначение пользователю роли Помощник")
+    @PreAuthorize("@roleSecurityExpression.canAssignAssistantRole(#eventId)")
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/controller/RoleController.java:153]: Стремно мы конечно придумали 3 привилегии на назначение помощника, организатора и огранизационной роли

Comment (IP-13) [reply src/main/java/org/itmo/eventapp/main/controller/RoleController.java:153]: Но ничего не поделать

Comment (rimnvd) [reply src/main/java/org/itmo/eventapp/main/controller/RoleController.java:153]: На назначение организатора точно должна быть отдельная, но в остальном да, странно. Изначально не было возможности создавать новые роли, все было логично, но как она появилась, начались проблемы с логикой и не только

Comment (rimnvd) [reply src/main/java/org/itmo/eventapp/main/controller/RoleController.java:153]: возможно, стоит переименовать привилегию в назначение созданной организационной роли, чтобы было понятнее

File: src/main/java/org/itmo/eventapp/main/service/EventRoleService.java [hunk 3]
```diff
@@ -65,23 +68,21 @@ public void revokeOrganizationalRole(Integer userId, Integer roleId, Integer eve
...
             }
             if (!role.getType().equals(RoleType.EVENT))
-                throw new ResponseStatusException(HttpStatus.CONFLICT, String.format(ExceptionConst.INVALID_ROLE_TYPE,
-                        "организационная"));
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
+                        ExceptionConst.INVALID_ROLE_TYPE.formatted("организационная"));
         }
-        var user = userService.findById(userId);
-        var event = eventFindById(eventId);
+        userService.findById(userId);
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/service/EventRoleService.java:74]: Это просто проверка, что существует?

Comment (rimnvd) [reply src/main/java/org/itmo/eventapp/main/service/EventRoleService.java:74]: да

File: src/main/java/org/itmo/eventapp/main/service/RoleService.java [hunk 4]
```diff
@@ -52,9 +61,15 @@ public Role editRole(Integer id, RoleRequest roleRequest) {
...
+        var privilegeType = roleRequest.isEvent() ? PrivilegeType.EVENT : PrivilegeType.SYSTEM;
         roleRequest.privileges().stream()
                 .map(privilegeService::findById)
-                .forEach(editedRole::addPrivilege);
+                .forEach(privilege -> {
+                    if (checkInvalidPrivilegeType(privilege, privilegeType)) {
+                        throw new ResponseStatusException(HttpStatus.BAD_REQUEST, ExceptionConst.INVALID_PRIVILEGE_TYPE);
+                    }
+                    editedRole.addPrivilege(privilege);
+                });
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/service/RoleService.java:72]: Это в маппер может вынести, удобнее получится

File: src/main/java/org/itmo/eventapp/main/controller/RoleController.java (hunk 2 above)
Comment (IP-13) [reply src/main/java/org/itmo/eventapp/main/controller/RoleController.java:153]: Да не, названия понятные, непонятно зачем там сложно

File: src/main/java/org/itmo/eventapp/main/service/RoleService.java (hunk 4 above)
Comment (rimnvd) [reply src/main/java/org/itmo/eventapp/main/service/RoleService.java:72]: Прямо в маппере добавлять привилегии, что ли?

Comment (IP-13) [reply src/main/java/org/itmo/eventapp/main/service/RoleService.java:72]: Да, они в запросе приходят, насколько я понял, так что можно в маппере сразу все делать, и функцию checkInvalidPrivilegeType в маппер вынести

**Comment by IP-13** [comment]
|Overall Project|75.32% **`-5.95%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|26.49%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[PrivilegeService.java](https://github.com/itmo-events-app/event-app/blob/02db341ad22ca0ce665e78eb4998dd3b3ee1d07d/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FPrivilegeService.java)|100%|:green_apple:|
|[UserService.java](https://github.com/itmo-events-app/event-app/blob/02db341ad22ca0ce665e78eb4998dd3b3ee1d07d/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FUserService.java)|89.02% **`-3.66%`**|:green_apple:|
|[RoleService.java](https://github.com/itmo-events-app/event-app/blob/02db341ad22ca0ce665e78eb4998dd3b3ee1d07d/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FRoleService.java)|67.01% **`-12.89%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/02db341ad22ca0ce665e78eb4998dd3b3ee1d07d/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|29.45% **`-31.49%`**|:x:|
|[RoleController.java](https://github.com/itmo-events-app/event-app/blob/02db341ad22ca0ce665e78eb4998dd3b3ee1d07d/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FRoleController.java)|11.79% **`-7.69%`**|:x:|
|[RoleSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/02db341ad22ca0ce665e78eb4998dd3b3ee1d07d/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FRoleSecurityExpression.java)|5.52% **`-94.48%`**|:x:|

**Pull request merged**